File: packages/gavio-java/gavio-core/src/test/java/io/gavio/platform/PlatformRuntimeTest.java
package io.gavio.platform;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import io.gavio.json.Json;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

class PlatformRuntimeTest {

    @Test
    @SuppressWarnings("unchecked")
    void buildsSharedPlatformRuntimeProfile() throws Exception {
        Map<String, Object> vector = loadVector();
        Map<String, Object> profile = buildFromVector((Map<String, Object>) vector.get("readyProfileInput"));

        assertEquals(vector.get("readyProfile"), profile);
        PlatformRuntimeVerification result = PlatformRuntime.verify(profile);
        assertTrue(result.valid());
        assertEquals(List.of(), result.errors());
        assertEquals(profile.get("profileHash"), result.computedHash());
    }

    @Test
    @SuppressWarnings("unchecked")
    void reportsPlatformRuntimeReadinessGaps() throws Exception {
        Map<String, Object> gapCase = (Map<String, Object>) loadVector().get("gapCase");
        Map<String, Object> profile = buildFromVector((Map<String, Object>) gapCase.get("input"));

        assertEquals(gapCase.get("expectedReadiness"), profile.get("readiness"));
        assertFalse((Boolean) ((Map<String, Object>) profile.get("readiness")).get("ready"));
    }

    @Test
    @SuppressWarnings("unchecked")
    void rejectsTamperedOrContentBearingProfile() throws Exception {
        Map<String, Object> profile = buildFromVector((Map<String, Object>) loadVector().get("readyProfileInput"));
        Map<String, Object> runtime = (Map<String, Object>) profile.get("runtime");
        runtime.put("eventExportMode", "full_local_debug");
        runtime.put("rawPrompt", "do not store me");

        PlatformRuntimeVerification result = PlatformRuntime.verify(profile);

        assertFalse(result.valid());
        assertTrue(result.errors().contains("profileHash does not match profile content"));
        assertTrue(result.errors().contains("profile contains content-bearing keys"));
        assertTrue(result.errors().contains("readiness does not match profile content"));
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> buildFromVector(Map<String, Object> input) {
        return PlatformRuntime.builder((String) input.get("profileId"))
                .generatedAt((String) input.get("generatedAt"))
                .sdk(
                        (String) ((Map<String, Object>) input.get("sdk")).get("name"),
                        (String) ((Map<String, Object>) input.get("sdk")).get("version"))
                .runtime((Map<String, Object>) input.get("runtime"))
                .surfaces((List<String>) input.get("surfaces"))
                .exporters((List<String>) input.get("exporters"))
                .integrations((List<String>) input.get("integrations"))
                .controls((List<Map<String, Object>>) input.get("controls"))
                .evidence((Map<String, Object>) input.get("evidence"))
                .build();
    }

    private static Map<String, Object> loadVector() throws Exception {
        return Json.parseObject(Files.readString(repoRoot().resolve("test-vectors/platform-runtime/profile.json")));
    }

    private static Path repoRoot() {
        Path cwd = Path.of("").toAbsolutePath();
        for (Path p = cwd; p != null; p = p.getParent()) {
            if (Files.isDirectory(p.resolve("test-vectors"))) {
                return p;
            }
        }
        throw new AssertionError("repository root not found from " + cwd);
    }
}

File: packages/gavio-js/tests/unit/platform-runtime.test.ts
import { describe, expect, it } from 'vitest'
import { readFileSync } from 'node:fs'
import { fileURLToPath } from 'node:url'
import {
  buildPlatformRuntimeProfile,
  verifyPlatformRuntimeProfile,
} from '../../src/platform-runtime.js'

const vector = JSON.parse(
  readFileSync(
    fileURLToPath(new URL('../../../../test-vectors/platform-runtime/profile.json', import.meta.url)),
    'utf8',
  ),
) as {
  readyProfileInput: Parameters<typeof buildPlatformRuntimeProfile>[0]
  readyProfile: Record<string, unknown>
  gapCase: {
    input: Parameters<typeof buildPlatformRuntimeProfile>[0]
    expectedReadiness: Record<string, unknown>
  }
}

describe('platform runtime profile', () => {
  it('builds the shared platform runtime profile', () => {
    const profile = buildPlatformRuntimeProfile(vector.readyProfileInput)

    expect(profile).toEqual(vector.readyProfile)
    const result = verifyPlatformRuntimeProfile(profile)
    expect(result.valid).toBe(true)
    expect(result.errors).toEqual([])
    expect(result.computedHash).toBe(profile.profileHash)
  })

  it('reports platform runtime readiness gaps', () => {
    const profile = buildPlatformRuntimeProfile(vector.gapCase.input)

    expect(profile.readiness).toEqual(vector.gapCase.expectedReadiness)
    expect(profile.readiness.ready).toBe(false)
  })

  it('rejects tampered or content-bearing profiles', () => {
    const profile = buildPlatformRuntimeProfile(vector.readyProfileInput) as Record<string, unknown>
    ;(profile.runtime as Record<string, unknown>).eventExportMode = 'full_local_debug'
    ;(profile.runtime as Record<string, unknown>).rawPrompt = 'do not store me'

    const result = verifyPlatformRuntimeProfile(profile)

    expect(result.valid).toBe(false)
    expect(result.errors).toContain('profileHash does not match profile content')
    expect(result.errors).toContain('profile contains content-bearing keys')
    expect(result.errors).toContain('readiness does not match profile content')
  })
})
